lib/libspl/backtrace.c

```diff
@@ -120,26 +136,42 @@ libspl_backtrace(int fd)
 	 * thing.
 	 */
 	uint_t cols = 0;
-	for (uint_t regnum = 0; regnum <= UNW_TDEP_LAST_REG; regnum++) {
+	for (uint_t regnum = 0; regnum <= LAST_REG_INDEX; regnum++) {
 		/*
 		 * Get the value. Any error probably means the register
-		 * doesn't exist, and we skip it.
+		 * doesn't exist, and we skip it. LLVM libunwind iterates over
+		 * fp registers in the same list, however they have to be
+		 * accessed using unw_get_fpreg instead. Here, we just ignore
+		 * them.
 		 */
+#if defined(LIBUNWIND_GCC)
 		if (unw_get_reg(&cp, regnum, &v) < 0)
 			continue;
+#elif defined(LIBUNWIND_LLVM)
+		if (unw_is_fpreg(&cp, regnum) ||
+		    unw_get_reg(&cp, regnum, &v) < 0)
+			continue;
+#endif
 
 		/*
-		 * Register name. If libunwind doesn't have a name for it,
+		 * Register name. If GCC libunwind doesn't have a name for it,
 		 * it will return "???". As a shortcut, we just treat '?'
-		 * is an alternate end-of-string character.
+		 * is an alternate end-of-string character. LLVM libunwind will
+		 * return the string 'unknown register', which we detect by
+		 * checking if the register name is longer than 5 characters.
 		 */
+#if defined(LIBUNWIND_GCC)
 		const char *name = unw_regname(regnum);
+#elif defined(LIBUNWIND_LLVM)
+		const char *name = unw_regname(&cp, regnum);
+#endif
 		for (n = 0; name[n] != '\0' && name[n] != '?'; n++) {}
-		if (n == 0) {
+		if (n == 0 || n > 5) {
```

Comment:
Are there any architectures with register names longer than five chars? :sweat_smile: I doubt it matters, we'll extend it if it ever comes up.

Reply:
Had a quick look - all the register names are here: https://github.com/llvm/llvm-project/blob/main/libunwind/src/Registers.hpp

Looks like there are a few 5 letter registers (xmm10 .. xmm15 on x86 and vlenb on on riscv) but nothing longer :)